File: internal/algorithms/redis/leaky_bucket.go
package redis

import (
	"context"
	"fmt"
	"time"

	"github.com/AliRizaAynaci/gorl/core"
	"github.com/redis/go-redis/v9"
)

type LeakyBucketLimiter struct {
	limit    int
	window   time.Duration
	client   *redis.Client
	prefix   string
	failOpen bool
}

var leakyBucketLua = `
	local key = KEYS[1]
	local now = tonumber(ARGV[1])
	local limit = tonumber(ARGV[2])
	local window = tonumber(ARGV[3])

	local bucket = redis.call("HMGET", key, "water", "last_leak")
	local water = tonumber(bucket[1]) or 0
	local last_leak = tonumber(bucket[2]) or now

	local elapsed = now - last_leak
	local leak_rate = limit / window
	local leaked = elapsed * leak_rate
	water = math.max(0, water - leaked)
	last_leak = now

	if water < limit then
		redis.call("HMSET", key, "water", water + 1, "last_leak", now)
		redis.call("EXPIRE", key, window)
		return 1
	else
		redis.call("HMSET", key, "water", water, "last_leak", now)
		redis.call("EXPIRE", key, window)
		return 0
	end
`

func (r *LeakyBucketLimiter) Allow(key string) (bool, error) {
	ctx := context.Background()
	now := time.Now().Unix()
	redisKey := fmt.Sprintf("%s:%s", r.prefix, key)
	result, err := r.client.Eval(ctx, leakyBucketLua, []string{redisKey},
		now, r.limit, int(r.window.Seconds()),
	).Result()
	if err != nil {
		// If the backend (Redis) is unavailable, either allow or block the request
		// depending on the failOpen policy. Always return a custom error for observability.
		if r.failOpen {
			return true, core.ErrBackendUnavailable
		}
		return false, core.ErrBackendUnavailable
	}
	allowed, ok := result.(int64)
	return ok && allowed == 1, nil
}

func NewLeakyBucketLimiter(cfg core.Config) core.Limiter {
	opt, _ := redis.ParseURL(cfg.RedisURL)
	client := redis.NewClient(opt)
	return &LeakyBucketLimiter{
		limit:    cfg.Limit,
		window:   cfg.Window,
		client:   client,
		prefix:   "gorl:lb",
		failOpen: cfg.FailOpen, // Pass config failOpen
	}
}

File: storage/redis/lua/leaky_bucket.lua
local limit = tonumber(ARGV[1])
local now_us = tonumber(ARGV[2])
local window_us = tonumber(ARGV[3])
local ttl_ms = tonumber(ARGV[4])

local us_per_token = math.floor(window_us / limit)
if us_per_token < 1 then
  us_per_token = 1
end

local water = tonumber(redis.call("GET", KEYS[1]) or "0")
local last_leak = tonumber(redis.call("GET", KEYS[2]) or "0")

if last_leak == 0 then
  water = 0
  last_leak = now_us
else
  local elapsed = now_us - last_leak
  local leaked = math.floor(elapsed / us_per_token)
  if leaked > 0 then
    water = water - leaked
    if water < 0 then
      water = 0
    end
    last_leak = last_leak + (leaked * us_per_token)
  end
end

local allowed = 0
if water < limit then
  allowed = 1
  water = water + 1
end

local elapsed_since_leak = now_us - last_leak
local next_leak = us_per_token - elapsed_since_leak
if next_leak < 0 then
  next_leak = 0
end

local reset_us = 0
if water > 0 then
  reset_us = (water * us_per_token) - elapsed_since_leak
  if reset_us < 0 then
    reset_us = 0
  end
end

local remaining = limit - water
if remaining < 0 then
  remaining = 0
end

redis.call("SET", KEYS[1], water, "PX", ttl_ms)
redis.call("SET", KEYS[2], last_leak, "PX", ttl_ms)

local retry_after_us = 0
if allowed == 0 then
  retry_after_us = next_leak
end

return {allowed, remaining, reset_us, retry_after_us}
